### DISCUSSION TITLE
How to support both old and new path / fspath ?

### DISCUSSION THREAD

**Discussion by vidartf** [pytest-dev/pytest#10261] [General] [answered]
I help maintain a pytest plugin, and I see there is a deprecation of the `path` argument to `pytest_collect_file` in favor of the new `file_path`. My first naive attempt to resolve this deprecation is to simply add the `file_path` argument to the function signature. This will prevent the deprecation warning on 7.x, but on pytest < 7.0.0, it raises an exception:

> Argument(s) {'file_path'} are declared in the hookimpl but can not be found in the hookspec

Is there any way to fix the deprecation warning without dropping support for pytest < 7.0.0 ? Or is this forced?

**Comment by nicoddemus** [comment, marked as answer]
Unfortunately at the moment there is no simple solution for that.

If you need to support old pytest versions and do not want to see the warning, I guess one workaround is to declare different hooks depending on the pytest version:

```python
def collect_file_hook(file_path: Path, parent: Collector) -> Collector:
    # The actual implementation goes here.
    ...

# Declare the hook depending on pytest's version.
version_tuple = getattr(pytest, "version_tuple", None)

if version_tuple and version_tuple[0] >= 7:
    # Modern version.
    def pytest_collect_file(file_path: Path, parent: Collector) -> Collector:
        return collect_file_hook(file_path, parent)

else:
    # Backward compatibility version.        
    def pytest_collect_file(path: py.path.local, parent: Collector) -> Collector:
        return collect_file_hook(Path(file_path), parent)
```

Definitely less than ideal, but that is what I can come up with for the moment.
